File: android-native/app/src/main/java/com/example/kpkn/screens/workout/WorkoutVisualModels.kt
package com.example.kpkn.screens.workout

import androidx.compose.ui.graphics.Color

internal enum class ExerciseDrainOverlayChannelV2 {
    ENERGY,
    BACK,
    MUSCLE,
}

internal data class ExerciseDrainOverlayItemV2(
    val label: String,
    val delta: Int,
    val channel: ExerciseDrainOverlayChannelV2,
)

internal data class ExerciseDrainOverlayStateV2(
    val key: String,
    val exerciseName: String,
    val items: List<ExerciseDrainOverlayItemV2>,
)

internal data class WorkoutStageTransitionTargetV2(
    val exerciseId: String,
    val order: Int,
    val label: String,
)

internal enum class WorkoutSetCardVisualState {
    FUTURE,
    COMPLETED,
    SKIPPED,
    ACTIVE,
}

internal data class WorkoutSetPagerItem(
    val index: Int,
    val label: String,
    val state: WorkoutSetCardVisualState,
    val isEditing: Boolean,
    val side: String? = null,
    val pulseToken: Long? = null,
    val isWarmupOrFeedback: Boolean = false,
)

internal val VolumeReplacedStepperTint = Color(0xFFE53935)

internal fun workoutSetPagerAccent(
    state: WorkoutSetCardVisualState,
    colorScheme: androidx.compose.material3.ColorScheme,
    isWarmupOrFeedback: Boolean = false,
    sessionAccentColor: Color? = null,
    volumeReplaced: Boolean = false,
): Color {
    val accent = if (volumeReplaced) {
        VolumeReplacedStepperTint
    } else {
        sessionAccentColor ?: colorScheme.primary
    }
    return when {
        isWarmupOrFeedback -> androidx.compose.ui.graphics.lerp(accent, Color.White, 0.22f)
        state == WorkoutSetCardVisualState.ACTIVE -> accent
        state == WorkoutSetCardVisualState.COMPLETED -> androidx.compose.ui.graphics.lerp(accent, Color.White, 0.18f)
        state == WorkoutSetCardVisualState.SKIPPED -> colorScheme.outline
        else -> colorScheme.onSurfaceVariant
    }
}

internal fun resolveWorkoutHeaderGroupLabel(
    partName: String?,
    type: String?,
    category: String?,
): String? {
    val explicitPart = partName
        ?.trim()
        ?.takeIf {
            it.isNotBlank() &&
                !it.equals("sesión", ignoreCase = true) &&
                !it.equals("sesion", ignoreCase = true) &&
                !it.equals("sesión principal", ignoreCase = true) &&
                !it.equals("sesion principal", ignoreCase = true)
        }
    return normalizeWorkoutHeaderLabel(explicitPart ?: type?.trim()?.takeIf { it.isNotBlank() } ?: category?.trim())
}

internal fun normalizeWorkoutHeaderLabel(raw: String?): String? {
    val value = raw?.trim()?.replace(Regex("\\s+"), " ") ?: return null
    if (value.isBlank()) return null
    val letterChars = value.filter { it.isLetter() }
    val upperRatio = if (letterChars.isEmpty()) {
        0.0
    } else {
        letterChars.count { it.isUpperCase() }.toDouble() / letterChars.length
    }
    // Preserve intentional ALL CAPS (and repair mixed leftovers like PRINCIPALEs from IME).
    if (upperRatio >= 0.75) return value.uppercase()
    if (value.equals("principales", ignoreCase = true)) return "Principales"
    if (value.equals("principal", ignoreCase = true)) return "Principal"
    return value
        .replace(Regex("principales", RegexOption.IGNORE_CASE), "Principales")
        .replace(Regex("principal", RegexOption.IGNORE_CASE), "Principal")
}

File: ios-native/KPKNFit/KPKNFit/Presentation/Screens/Workout/WorkoutVisualModels.swift
import SwiftUI

internal enum ExerciseDrainOverlayChannelV2: String, Codable {
    case ENERGY
    case BACK
    case MUSCLE
}

internal struct ExerciseDrainOverlayItemV2: Codable {
    let label: String
    let delta: Int
    let channel: ExerciseDrainOverlayChannelV2
}

internal struct ExerciseDrainOverlayStateV2: Codable {
    let key: String
    let exerciseName: String
    let items: [ExerciseDrainOverlayItemV2]
}

internal struct WorkoutStageTransitionTargetV2: Codable {
    let exerciseId: String
    let order: Int
    let label: String
}

internal enum WorkoutSetCardVisualState: String, Codable {
    case FUTURE
    case COMPLETED
    case SKIPPED
    case ACTIVE
}

internal struct WorkoutSetPagerItem: Codable {
    let index: Int
    let label: String
    let state: WorkoutSetCardVisualState
    let isEditing: Bool
    let side: String?
    let pulseToken: Int64?
    let isWarmupOrFeedback: Bool

    init(
        index: Int,
        label: String,
        state: WorkoutSetCardVisualState,
        isEditing: Bool,
        side: String? = nil,
        pulseToken: Int64? = nil,
        isWarmupOrFeedback: Bool = false
    ) {
        self.index = index
        self.label = label
        self.state = state
        self.isEditing = isEditing
        self.side = side
        self.pulseToken = pulseToken
        self.isWarmupOrFeedback = isWarmupOrFeedback
    }
}

internal let WORKOUT_WARMUP_BLUE = Color(red: 68 / 255.0, green: 138 / 255.0, blue: 255 / 255.0)

internal func workoutSetPagerAccent(
    state: WorkoutSetCardVisualState,
    primaryColor: Color = .accentColor,
    tertiaryColor: Color = .orange,
    outlineColor: Color = .gray,
    futureColor: Color = .secondary,
    isWarmupOrFeedback: Bool = false,
    sessionAccentColor: Color? = nil
) -> Color {
    if isWarmupOrFeedback {
        return WORKOUT_WARMUP_BLUE
    }
    if let sessionAccentColor = sessionAccentColor {
        return sessionAccentColor
    }
    switch state {
    case .ACTIVE: return primaryColor
    case .COMPLETED: return tertiaryColor
    case .SKIPPED: return outlineColor
    case .FUTURE: return futureColor
    }
}

internal func resolveWorkoutHeaderGroupLabel(
    partName: String?,
    type: String?,
    category: String?
) -> String? {
    let explicitPart: String?
    if let part = partName?.trimmingCharacters(in: .whitespacesAndNewlines),
       !part.isEmpty,
       part.caseInsensitiveCompare("sesión") == .orderedSame ||
       part.caseInsensitiveCompare("sesion") == .orderedSame ||
       part.caseInsensitiveCompare("sesión principal") == .orderedSame ||
       part.caseInsensitiveCompare("sesion principal") == .orderedSame {
        explicitPart = nil
    } else {
        explicitPart = partName?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false ? partName?.trimmingCharacters(in: .whitespacesAndNewlines) : nil
    }

    let cleanType = type?.trimmingCharacters(in: .whitespacesAndNewlines)
    let explicitType = (cleanType?.isEmpty == false) ? cleanType : nil

    let cleanCategory = category?.trimmingCharacters(in: .whitespacesAndNewlines)
    let explicitCategory = (cleanCategory?.isEmpty == false) ? cleanCategory : nil

    return normalizeWorkoutHeaderLabel(explicitPart ?? explicitType ?? explicitCategory)
}

private func normalizeWorkoutHeaderLabel(_ raw: String?) -> String? {
    guard let trimmed = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !trimmed.isEmpty else { return nil }
    
    // Replace multiple spaces with a single space
    let components = trimmed.components(separatedBy: .whitespacesAndNewlines).filter { !$0.isEmpty }
    let value = components.joined(separator: " ")
    
    if value.caseInsensitiveCompare("principales") == .orderedSame {
        return "Principales"
    }
    if value.caseInsensitiveCompare("principal") == .orderedSame {
        return "Principal"
    }
    
    let letterChars = value.filter { $0.isLetter }
    let upperCount = letterChars.filter { $0.isUppercase }.count
    let upperRatio = letterChars.isEmpty ? 0.0 : Double(upperCount) / Double(letterChars.count)
    
    if upperRatio >= 0.75 {
        return value.uppercased()
    }
    
    // Case-insensitive replacements for "principales" and "principal"
    var replaced = value
    if let regexPrincipales = try? NSRegularExpression(pattern: "principales", options: .caseInsensitive) {
        replaced = regexPrincipales.stringByReplacingMatches(in: replaced, options: [], range: NSRange(replaced.startIndex..., in: replaced), withTemplate: "Principales")
    }
    if let regexPrincipal = try? NSRegularExpression(pattern: "principal", options: .caseInsensitive) {
        replaced = regexPrincipal.stringByReplacingMatches(in: replaced, options: [], range: NSRange(replaced.startIndex..., in: replaced), withTemplate: "Principal")
    }
    
    return replaced
}
